sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -524,10 +543,7 @@ case class HashShuffleSpec(
     // will add shuffles with the default partitioning of `ClusteredDistribution`, which uses all
     // the join keys.
     if (SQLConf.get.getConf(SQLConf.REQUIRE_ALL_CLUSTER_KEYS_FOR_CO_PARTITION)) {
-      partitioning.expressions.length == distribution.clustering.length &&
-        partitioning.expressions.zip(distribution.clustering).forall {
-          case (l, r) => l.semanticEquals(r)
-        }
+      partitioning.isPartitionedOnFullKeys(distribution)
```

Comment:
Although it is beyond the scope of the PR, same thing applies here. Would we need to require strict order of keys?

Reply:
I think we probably need strict order here, as it feels more safer with requiring of order, and this is the original behavior when join requires `HashClusteredDistribution`. This essentially restores the join's behavior to what it was with `HashClusteredDistribution`.